packaging/elasticsearch_plugin.py

```diff
@@ -0,0 +1,162 @@
+"""
+
+DOCUMENTATION = '''
+---
+module: elasticsearch_plugin
+short_description: Manage Elasticsearch plugins
+description:
+    - Manages Elasticsearch plugins.
+version_added: ""
+author: Mathew Davies (@ThePixelDeveloper)
+options:
+    name:
+        description:
+            - Name of the plugin to install
+        required: True
+    state:
+        description:
+            - Desired state of a plugin.
+        required: False
+        choices: [present, absent]
+        default: present
+    url:
+        description:
+            - Set exact URL to download the plugin from
+        required: False
+        default: None
+    timeout:
+        description:
+            - Timeout setting: 30s, 1m, 1h... (1m by default)
+        required: False
+    plugin_bin:
+        description:
+            - Location of the plugin binary
+        required: False
+        default: /usr/share/elasticsearch/bin/plugin
+    plugin_dir:
+        description:
+            - Your configured plugin directory specified in Elasticsearch
+        required: False
+        default: /usr/share/elasticsearch/plugins/
+    version:
+        description: 
+            - Version of the plugin to be installed.
+        default: None
+'''
+
+EXAMPLES = '''
+# Install Elasticsearch head plugin
+- elasticsearch_plugin: state=present name="mobz/elasticsearch-head"
+'''
+
+
+def parse_plugin_repo(string):
+    elements = string.split("/")
+
+    # We first consider the simplest form: pluginname
+    repo = elements[0]
+
+    # We consider the form: username/pluginname
+    if len(elements) > 1:
+        repo = elements[1]
+
+    # remove elasticsearch- prefix
+    # remove es- prefix
+    for string in ("elasticsearch-", "es-"):
+        if repo.startswith(string):
+            return repo[len(string):]
+
+    return repo
+
+
+def is_plugin_present(plugin_dir, working_dir):
+    return os.path.isdir(os.path.join(working_dir, plugin_dir))
+
+
+def parse_error(string):
+    reason = "reason: "
+    return string[string.index(reason) + len(reason):].strip()
+
+
+def main():
+
+    package_state_map = dict(
+        present="--install",
+        absent="--remove"
+    )
+
+    module = AnsibleModule(
+        argument_spec=dict(
+            name=dict(required=True),
+            state=dict(default="present", choices=package_state_map.keys()),
+            url=dict(default=None),
+            timeout=dict(default="1m"),
+            plugin_bin=dict(default="/usr/share/elasticsearch/bin/plugin"),
+            plugin_dir=dict(default="/usr/share/elasticsearch/plugins/"),
+            version=dict(default=None)
+        )
+    )
+
+    plugin_bin = module.params["plugin_bin"]
+    plugin_dir = module.params["plugin_dir"]
+    name = module.params["name"]
+    state = module.params["state"]
+    url = module.params["url"]
+    timeout = module.params["timeout"]
+    version = module.params["version"]
+
+    present = is_plugin_present(parse_plugin_repo(name), plugin_dir)
```

Comment:
Unfortunately ES doesn't store versions installed. It just extracts the plugin into a directory. The plugin tool doesn't specify versions either.

```
➜  ~elasticsearch  ./bin/plugin -l
Picked up JAVA_TOOL_OPTIONS: -javaagent:/usr/share/java/jayatanaag.jar 
Picked up _JAVA_OPTIONS: -Dawt.useSystemAAFontSettings=lcd                       -Dsun.java2d.xrender=true
Installed plugins:
    - head
    - bigdesk
```